mlflow/utils/virtualenv.py

```diff
@@ -0,0 +1,273 @@
+    for deps in filter(None, [python_env.build_dependencies, python_env.dependencies]):
+        # Use a unique name to avoid conflicting custom requirements files logged by a user
+        with tempfile.TemporaryDirectory() as tmpdir:
+            # Create a temporary requirements file
+            tmpdir = Path(tmpdir)
+            tmp_req_file = tmpdir.joinpath(f"requirements.{uuid.uuid4().hex}.txt")
+            tmp_req_file.write_text("\n".join(deps))
+            # Copy `requirements.txt` and `constraints.txt` in the temporary directory.
+            # --- Why do we need to copy them? ---
+            # Let's say `tmp_req_file` is stored in /tmp/123 and it contains '-r requirements.txt'.
+            # pip resolves this reference to '-r /tmp/123/requirements.txt'. If we didn't copy
+            # requirements.txt in /tmp/123, pip would complain /tmp/123/requirements.txt is not
+            # found. The same thing applies to constraints.txt.
+            req_file = local_model_path.joinpath(_REQUIREMENTS_FILE_NAME)
+            if req_file.exists():
+                shutil.copy(req_file, tmpdir)
+            con_file = local_model_path.joinpath(_CONSTRAINTS_FILE_NAME)
+            if con_file.exists():
+                shutil.copy(con_file, tmpdir)
+            # In windows `pip install pip==x.y.z` causes the following error:
+            # `[WinError 5] Access is denied: 'C:\path\to\pip.exe`
+            # This can be avoided by using `python -m`.
+            cmd = _join_commands(activate_cmd, f"python -m pip install -r {tmp_req_file}")
+            _exec_cmd(cmd, capture_output=False)
+
+    return activate_cmd
+
+
+def _get_or_create_virtualenv(local_model_path, env_id=None):
+    """
+    Restores an MLflow model's environment with pyenv and virtualenv and returns a command
+    to activate it.
+
+    :param local_model_path: Local directory containing the model artifacts.
+    :param env_id: Optional string that is added to the contents of the yaml file before
+                   calculating the hash. It can be used to distinguish environments that have the
+                   same conda dependencies but are supposed to be different based on the context.
+                   For example, when serving the model we may install additional dependencies to the
+                   environment after the environment has been activated.
+    :return: Command to activate the created virtualenv environment
+             (e.g. "source /path/to/bin/activate").
+    """
+    _validate_pyenv_is_available()
+    _validate_virtualenv_is_available()
+
+    # Read environment information
+    local_model_path = Path(local_model_path)
+    python_env = _get_python_env(local_model_path)
+
+    # Create an environment
+    python_bin_path = _install_python(python_env.python)
+    env_root = Path(_get_mlflow_virtualenv_root())
+    env_root.mkdir(parents=True, exist_ok=True)
+    env_name = _get_mlflow_env_name(str(python_env) + (env_id or ""))
```

Comment:
This env name remains the same even if I change the contents of `requirements.txt` because `str(python_env)` is:

```
PYTHON ENV {'python': '3.8.12', 'build_dependencies': ['pip==21.2.4', 'setuptools==58.0.4', 'wheel==0.37.1'], 'dependencies': ['-r requirements.txt']}
```

I tried manually changing a model requirement in `requirements.txt` and then serving the model:

```
2022/04/12 13:01:05 INFO mlflow.utils.virtualenv: Environment /Users/corey.zumar/.mlflow/envs/mlflow-faac9d7af721b9a11af03a0dcc96d30b5f28bb05 already exists
...
```

Can we account for changes in `requirements.txt` and add test coverage?

Reply:
Thanks for the catch, I'll fix this.

Reply:
```python
# doesn't work
env_name = _get_mlflow_env_name(str(python_env) + open(requirements.txt).read() + (env_id or ""))
```

Reply:
Offline discussion: Use `_parse_pip_requirements` and create hash from the return value.